app/src/main/java/ch/sheremet/dasfoo/rover/android/client/MainActivity.java

```diff
@@ -5,64 +5,78 @@
 import android.widget.Toast;
 
-import java.util.concurrent.TimeUnit;
+import com.google.android.gms.security.ProviderInstaller;
 
+import ch.sheremet.dasfoo.rover.android.client.grpc.authentication.GrpcProviderInstaller;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.AbstractGrpcTaskExecutor;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.EncodersReadingTask;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.GettingBoardInfoTask;
-import ch.sheremet.dasfoo.rover.android.client.grpc.task.IOnGrpcTaskCompleted;
+import ch.sheremet.dasfoo.rover.android.client.grpc.task.GrpcConnection;
 import ch.sheremet.dasfoo.rover.android.client.grpc.task.MovingRoverTask;
-import dasfoo.grpc.roverserver.nano.RoverServiceGrpc;
-import io.grpc.ManagedChannel;
-import io.grpc.ManagedChannelBuilder;
 
-public class MainActivity extends AppCompatActivity implements IOnGrpcTaskCompleted {
+public class MainActivity extends AppCompatActivity {
 
     private Button mMoveForwardButton;
     private Button mInfoButton;
     private Button mReadEncodersButton;
     private EditText mHostEdit;
     private EditText mPortEdit;
     private TextView mResultText;
+    private GrpcConnection mGrpcConnection;
+    private static final String TAG = MainActivity.class.getName();
 
     @Override
-    protected void onCreate(Bundle savedInstanceState) {
+    protected void onCreate(final Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
         mMoveForwardButton = (Button) findViewById(R.id.move_forward_button);
-        mMoveForwardButton.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                moveForward();
-            }
-        });
+        mMoveForwardButton.setOnClickListener(onClickListener);
         mInfoButton = (Button) findViewById(R.id.info_button);
-        mInfoButton.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                getInfo();
-            }
-        });
+        mInfoButton.setOnClickListener(onClickListener);
         mReadEncodersButton = (Button) findViewById(R.id.read_encoders_button);
-        mReadEncodersButton.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                readEncoders();
-            }
-        });
+        mReadEncodersButton.setOnClickListener(onClickListener);
         mHostEdit = (EditText) findViewById(R.id.host_edit_text);
         mPortEdit = (EditText) findViewById(R.id.port_edit_text);
         mResultText = (TextView) findViewById(R.id.grpc_response_text);
+
+        // Android relies on a security Provider to provide secure network communications.
+        // It verifies that the security provider is up-to-date.
+        ProviderInstaller.installIfNeededAsync(this, new GrpcProviderInstaller());
```

Comment:
Is it possible to use `installIfNeeded` instead?  Looks like it might not need any installListener.